File: dotnet-aspnetcore-js/OAuth.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace HostedFields;

/// <summary>
/// OAuth 1.0a RSA-SHA256 request signing.
/// <see href="https://doc.payneteasy.com/integration/general_api_usage/request_authentication_methods/oauth.html"/>
/// </summary>
public static class OAuth
{
    /// <summary>
    /// The Authorization header for a signed API call. <paramref name="parameters"/> are the
    /// x-www-form-urlencoded parameters of the request, if any — the ephemeral ticket call sends
    /// none, and then only the oauth_* values are signed.
    /// </summary>
    public static string Header(
        string method,
        string endpoint,
        IReadOnlyDictionary<string, string> parameters,
        string consumerKey,
        RSA privateKey)
    {
        var oauth = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["oauth_consumer_key"] = consumerKey,
            ["oauth_nonce"] = Convert.ToHexString(RandomNumberGenerator.GetBytes(16)).ToLowerInvariant(),
            ["oauth_signature_method"] = "RSA-SHA256",
            ["oauth_timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture),
            ["oauth_version"] = "1.0",
        };

        var signed = new Dictionary<string, string>(parameters, StringComparer.Ordinal);
        foreach ((string name, string value) in oauth)
        {
            signed[name] = value;
        }

        byte[] signature = privateKey.SignData(
            Encoding.UTF8.GetBytes(BaseString(method, endpoint, signed)),
            HashAlgorithmName.SHA256,
            RSASignaturePadding.Pkcs1);
        oauth["oauth_signature"] = Convert.ToBase64String(signature);

        // Header values are not encoded by the transport, so they are encoded here
        return "OAuth " + string.Join(
            ", ",
            oauth.OrderBy(pair => pair.Key, StringComparer.Ordinal)
                 .Select(pair => $"{pair.Key}=\"{Encode(pair.Value)}\""));
    }

    /// <summary>
    /// The signature base string: METHOD&amp;url&amp;sorted-parameters, each part percent-encoded.
    /// </summary>
    public static string BaseString(string method, string endpoint, IReadOnlyDictionary<string, string> signed) =>
        method.ToUpperInvariant() + "&" + Encode(endpoint) + "&" + Encode(Normalize(signed));

    /// <summary>The signed parameters as <c>key=value</c> pairs, sorted by key and encoded.</summary>
    public static string Normalize(IReadOnlyDictionary<string, string> parameters) =>
        string.Join(
            "&",
            parameters.OrderBy(pair => pair.Key, StringComparer.Ordinal)
                      .Select(pair => Encode(pair.Key) + "=" + Encode(pair.Value)));

    /// <summary>
    /// RFC 3986 percent encoding, written out rather than taken from the framework.
    /// </summary>
    /// <remarks>
    /// This is not a form encoder and must never be replaced by one: form encoding writes a space
    /// as <c>+</c> and leaves <c>!'()*</c> alone, and either produces a signature the gateway
    /// rejects with a 401 that says nothing about why. The Sale <em>body</em> really is form
    /// encoded — that is what <see cref="Paynet"/>'s FormUrlEncodedContent is for — and the two
    /// must not be swapped.
    /// </remarks>
    public static string Encode(string value)
    {
        const string Unreserved = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";

        var encoded = new StringBuilder(value.Length);
        foreach (byte b in Encoding.UTF8.GetBytes(value))
        {
            if (Unreserved.IndexOf((char)b, StringComparison.Ordinal) >= 0)
            {
                encoded.Append((char)b);
            }
            else
            {
                encoded.Append(CultureInfo.InvariantCulture, $"%{b:X2}");
            }
        }
        return encoded.ToString();
    }
}

File: nodejs-express-ts-js/src/oauth.ts
// OAuth 1.0a RSA-SHA256 request signing.
// https://doc.payneteasy.com/integration/general_api_usage/request_authentication_methods/oauth.html

import { createSign, randomBytes } from 'node:crypto';
import { MERCHANT_LOGIN, PRIVATE_KEY } from './config.ts';

// RFC 3986 percent encoding: encodeURIComponent leaves ! ' ( ) * unescaped
export function encode(value: string): string {
  return encodeURIComponent(value).replace(/[!'()*]/g, (c) => `%${c.charCodeAt(0).toString(16).toUpperCase()}`);
}

// Accepts a PEM with real newlines or with escaped \n, and tolerates indentation
function normalizeKey(pem: string): string {
  return pem
    .replace(/\\n/g, '\n')
    .split('\n')
    .map((line) => line.trim())
    .join('\n');
}

/**
 * The signature base string: METHOD&url&sorted-parameters, each part percent-encoded. It carries
 * no secret, which is what makes it testable on its own — and it is where a signature usually
 * goes wrong, because every part has to be encoded to RFC 3986 rather than to the looser rules
 * the standard library applies.
 */
export function baseString(method: string, url: string, params: Record<string, string>): string {
  const normalized = Object.keys(params)
    .sort()
    .map((key) => `${encode(key)}=${encode(params[key] ?? '')}`)
    .join('&');
  return `${method.toUpperCase()}&${encode(url)}&${encode(normalized)}`;
}

/**
 * Builds the Authorization header for a signed API call.
 * bodyParams are the x-www-form-urlencoded parameters of the request, if any.
 */
export function buildAuthHeader(method: string, url: string, bodyParams: Record<string, string> = {}): string {
  const oauthParams: Record<string, string> = {
    oauth_consumer_key: MERCHANT_LOGIN,
    oauth_nonce: randomBytes(16).toString('hex'),
    oauth_signature_method: 'RSA-SHA256',
    oauth_timestamp: String(Math.floor(Date.now() / 1000)),
    oauth_version: '1.0',
  };

  const base = baseString(method, url, { ...bodyParams, ...oauthParams });

  const signature = createSign('RSA-SHA256').update(base).sign(normalizeKey(PRIVATE_KEY), 'base64');

  // Header values are not encoded by the transport, so encode them here
  const header = { ...oauthParams, oauth_signature: signature };
  return `OAuth ${Object.entries(header)
    .map(([key, value]) => `${key}="${encode(value)}"`)
    .join(', ')}`;
}
